src/vs/workbench/contrib/chat/browser/chat.ts

```diff
@@ -134,3 +138,23 @@ export interface IChatWidget {
 export interface IChatViewPane {
 	clear(): void;
 }
+
+
+export interface ICodeBlockActionContextProvider {
+	getCodeBlockContext(editor?: ICodeEditor): ICodeBlockActionContext | undefined;
+}
+export interface ICodeBlockContextProviderRegistry {
+	serviceBrand: undefined;
+}
+export class CodeBlockContextProviderRegistry implements ICodeBlockContextProviderRegistry {
+	serviceBrand: undefined;
+	static readonly _providers = new Map<string, ICodeBlockActionContextProvider>();
+	static get providers(): ICodeBlockActionContextProvider[] {
+		return [...this._providers.values()];
```

Comment:
Services shouldn't be singletons with static functions that others directly depend on. That defeats the purpose of injection. 

Using a service is preferred but so that it has just instance methods and no public static singleton instance.